File: 25.k-个一组翻转链表.java
/*
 * @lc app=leetcode.cn id=25 lang=java
 *
 * [25] K 个一组翻转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode nextHead = head;
        for (int i = 0; i < k; i++) {
            if (nextHead == null) return head;
            nextHead = nextHead.next;
        }

        ListNode next = reverseKGroup(nextHead, k);
        ListNode cur = head;
        for (int i = 0; i < k; i++) {
            ListNode temp = cur;
            cur = cur.next;
            temp.next = next;
            next = temp;
        }

        return next;
    }
}
// @lc code=end

File: 25.k-个一组翻转链表.py
#
# @lc app=leetcode.cn id=25 lang=python3
#
# [25] K 个一组翻转链表
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        if not head:
            return None
        dummy = ListNode()
        dummy.next = head
        tail = head
        p = head.next
        head.next = None
        count = 1
        while count < k and p:
            temp = p.next
            p.next = dummy.next
            dummy.next = p
            p = temp
            count += 1

        if count == k:
            tail.next = self.reverseKGroup(p, k)
            return dummy.next
        else:
            dummy2 = ListNode()
            p = dummy.next
            while p:
                temp = p.next
                p.next = dummy2.next
                dummy2.next = p
                p = temp
            return dummy2.next



        
# @lc code=end

